pkg/kopia/cli/internal/log/log.go

```diff
@@ -43,3 +44,31 @@ func (NopLogger) WithContext(ctx context.Context) log.Logger {
 func (NopLogger) WithError(err error) log.Logger {
 	return &NopLogger{}
 }
+
+// StringLogger is a logger that stores log messages in a slice of strings.
+type StringLogger []string
+
+func (l *StringLogger) Print(msg string, fields ...field.M) {
+	*l = append(*l, msg)
+}
+
+func (l *StringLogger) PrintTo(w io.Writer, msg string, fields ...field.M) {
+	*l = append(*l, msg)
+}
+
+func (l *StringLogger) WithContext(ctx context.Context) log.Logger {
+	return l
+}
+
+func (l *StringLogger) WithError(err error) log.Logger {
+	return l
+}
+
+func (l *StringLogger) MatchString(pattern string) bool {
+	for _, line := range *l {
+		if found, _ := regexp.MatchString(pattern, line); found {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
Should we move this logger related changes to a different PR since its not related to s2 storage flags functionality

Reply:
I included the `StringLogger` here primarily because it's key for testing and debugging S3 storage flags. 
We already have 12 PRs in our PR train, adding another one just for the logger could complicate things further. 

Keeping related changes together, like this logger that directly supports S3 storage flags, allows for smoother integration and testing. It seems more efficient to manage this way. 

Happy to discuss this further if there are any concerns!